File: guessing-game-bdd-atomic/src/main/java/tw/waterballsa/bddworkshop/service/GreetingService.java
package tw.waterballsa.bddworkshop.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import tw.waterballsa.bddworkshop.entity.Greeting;
import tw.waterballsa.bddworkshop.exceptions.BadRequestException;
import tw.waterballsa.bddworkshop.exceptions.RateLimitExceededException;
import tw.waterballsa.bddworkshop.repository.GreetingRepository;

import java.time.LocalDateTime;
import java.util.UUID;

@Service
public class GreetingService {
    private static final Logger logger = LoggerFactory.getLogger(GreetingService.class);
    private static final int MAX_GREETINGS_PER_MINUTE = 2;
    
    @Autowired
    private GreetingRepository greetingRepository;

    public String greet(String name) {
        // Check rate limit - count greetings in the last minute
        LocalDateTime oneMinuteAgo = LocalDateTime.now().minusMinutes(1);
        long greetingCount = greetingRepository.countGreetingsSince(oneMinuteAgo);
        
        logger.info("Current greeting count in last minute: {}", greetingCount);
        
        if (greetingCount >= MAX_GREETINGS_PER_MINUTE) {
            logger.warn("Rate limit exceeded. Current count: {}, Max allowed: {}", greetingCount, MAX_GREETINGS_PER_MINUTE);
            throw new RateLimitExceededException("Rate limit exceeded");
        }
        
        String message = String.format("Hello world, %s.", name);
        Greeting greeting = new Greeting(UUID.randomUUID().toString(), message, name, LocalDateTime.now());
        greetingRepository.save(greeting);
        
        logger.info("Greeting created successfully for name: {}", name);
        return message;
    }

    public void deleteGreetingByContent(String content) {
        logger.info("Attempting to delete greeting with content: {}", content);
        
        // Find the greeting by content
        Greeting greeting = greetingRepository.findByContent(content);
        
        if (greeting == null) {
            logger.warn("Greeting with content '{}' not found", content);
            throw new BadRequestException("Greeting with content '" + content + "' not found");
        }
        
        // Delete the greeting
        greetingRepository.delete(greeting);
        logger.info("Greeting with content '{}' deleted successfully", content);
    }
}

File: guessing-game-bdd-atomic/src/greeting/greeting.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, MoreThan } from 'typeorm';
import { v4 as uuidv4 } from 'uuid';
import { Greeting } from './entities/greeting.entity';
import { BadRequestException } from '../exceptions/bad-request.exception';
import { RateLimitExceededException } from '../exceptions/rate-limit-exceeded.exception';

@Injectable()
export class GreetingService {
  private readonly logger = new Logger(GreetingService.name);
  private static readonly MAX_GREETINGS_PER_MINUTE = 2;

  constructor(
    @InjectRepository(Greeting)
    private readonly greetingRepository: Repository<Greeting>,
  ) {}

  async greet(name: string): Promise<string> {
    // Check rate limit - count greetings in the last minute
    const oneMinuteAgo = new Date(Date.now() - 60000);
    const greetingCount = await this.greetingRepository.count({
      where: {
        createdAt: MoreThan(oneMinuteAgo),
      },
    });

    this.logger.log(`Current greeting count in last minute: ${greetingCount}`);

    if (greetingCount >= GreetingService.MAX_GREETINGS_PER_MINUTE) {
      this.logger.warn(`Rate limit exceeded. Current count: ${greetingCount}, Max allowed: ${GreetingService.MAX_GREETINGS_PER_MINUTE}`);
      throw new RateLimitExceededException('Rate limit exceeded');
    }

    const message = `Hello world, ${name}.`;
    const greeting = new Greeting(uuidv4(), message, name, new Date());
    await this.greetingRepository.save(greeting);

    this.logger.log(`Greeting created successfully for name: ${name}`);
    return message;
  }

  async deleteGreetingByContent(content: string): Promise<void> {
    this.logger.log(`Attempting to delete greeting with content: ${content}`);

    // Find the greeting by content
    const greeting = await this.greetingRepository.findOne({
      where: { content },
    });

    if (!greeting) {
      this.logger.warn(`Greeting with content '${content}' not found`);
      throw new BadRequestException(`Greeting with content '${content}' not found`);
    }

    // Delete the greeting
    await this.greetingRepository.remove(greeting);
    this.logger.log(`Greeting with content '${content}' deleted successfully`);
  }
}